### DISCUSSION TITLE
Prevent multiple logs of the same db query in Next.js

### DISCUSSION THREAD

**Discussion by Gejsi** [prisma/prisma#15043] [Q&A] [answered]
## Background
After following the best practice for creating a [prisma client in Next.js](https://www.prisma.io/docs/guides/database/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices), I encountered a problem with the [logging middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware/logging-middleware).
Suppose the `client` looks something like:
```ts
import { PrismaClient } from '@prisma/client'
import { env } from '../../env/server.mjs'

declare global {
  // eslint-disable-next-line no-var
  var prisma: PrismaClient | undefined
}

export const prisma =
  global.prisma ||
  new PrismaClient({
    log: ['query'],
  })

if (env.NODE_ENV !== 'production') {
  prisma.$use(async (params, next) => {
    const before = Date.now()
    await next(params)
    const after = Date.now()
    console.log(
      `Query ${params.model}.${params.action} took ${after - before}ms`
    )
  })

  global.prisma = prisma
}
```
An example prisma query would look something like:
![image](https://user-images.githubusercontent.com/25225304/187036007-34621c11-68a4-482a-acad-7b0fc98f7c83.png)
(which is correct)

## The problem
Due to hot reloading the query becomes something like:
```
prisma:query [...]
Query Exampe.findMany took 21ms
Query Exampe.findMany took 21ms
```
After more hot reloads the count of the same logs goes up everytime (not really, it's not very deterministic), and this can become pretty annoying quickly and maybe even confuse the developer into thinking that there are multiple queries.

Is there a clean solution to this problem?

## Video example
Simple steps to reproduce the middleware running multiple times:
1. start a Next.js app
2. fetch (through `curl` or something else) the `http://localhost:3000/api/examples/` route which has a simple db query.
3. cause a hot-reload (e.g. change homepage title)
4. re-fetch (step 2)

The `examples` api route is very simple:
```ts
const examples = async (req: NextApiRequest, res: NextApiResponse) => {
  const examples = await prisma.example.findMany()
  res.status(200).json(examples)
}
```

In the terminal (center), you can notice the query count increasing after the fetches.

https://user-images.githubusercontent.com/25225304/187036444-8bc9913c-4f3b-470c-81af-f5e80693f0cc.mp4

**Comment by pantharshit00** [comment, marked as answer]
Can you please also for client instance existing while creating the middleware? 
```ts
if (env.NODE_ENV !== 'production' && !global.prisma) {
```

**Reply by Gejsi** [reply]
That seems to be working. Thanks.
I'd suggest adding it to the docs :+1:
